### DISCUSSION TITLE
Initializing the Model in PySWMM

### DISCUSSION THREAD

**Discussion by Dinho4** [pyswmm/pyswmm#320] [Q&A]
The following is what view whenever I try to Initialize simulation, and don't know how to proceed.

from pyswmm import Simulation, Links, Subcatchments, Nodes, SystemStats

root = 'C:/Users/Malesela/Downloads/'
inpSWMM = 'CIV5000W Simulation Project.inp'

sim = Simulation(root + inpSWMM)

Cannot open input file C:/Users/Malesela/Downloads/CIV5000W Simulation Project.inpTraceback (most recent call last):
File "", line 1, in
File "/root/tutorial_env/lib/python3.9/site-packages/pyswmm/simulation.py", line 69, in init
self._model.swmm_open()
File "/root/tutorial_env/lib/python3.9/site-packages/pyswmm/swmm5.py", line 217, in swmm_open
solver.swmm_open(inpfile, rptfile, binfile)
File "/root/tutorial_env/lib/python3.9/site-packages/swmm/toolkit/solver.py", line 509, in swmm_open
return _solver.swmm_open(f1, f2, f3)
Exception:
ERROR 303: cannot open input file.

**Comment by jennwuu** [comment]
You are seeing this error because pyswmm cannot find the inp file that you provided in the path. 

Can you confirm if path (root + inpSWMM) is correct or if pyswmm installation has access to it?

**Reply by Dinho4** [reply]
I get the following when I try to confirm path:
>>> path (root + inpSWMM)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'path' is not defined

The File is saved under the following on my PC:
This PC>Downloads
So is my root = 'This PC/Downloads/' ?

The file name or format is:
CIV5000W Simulation Project.inp as an INP File
So is my InpSWMM = 'CIV5000W Simulation Project.inp' ?

Because if so, it still gives me the same problem as mentioned earlier.

And, how do you confirm/know if pyswmm installation has access to it?

**Reply by jennwuu** [reply]
What is your inp file path? 

Can you share the entire file path?

**Reply by jennwuu** [reply]
Is it your inp file C:/Users/Malesela/Downloads/CIV5000W Simulation Project.inp?

If so, try the following code snippet to run your simulation model.


```
from pyswmm import Simulation

with Simulation(r'C:/Users/Malesela/Downloads/CIV5000W Simulation Project.inp') as sim:
    for step in sim:
        print(sim.current_time)
```

You can check out this tutorial: https://medium.com/@wuu.jennifer/pyswmm-tutorial-how-to-install-and-to-start-using-pyswmm-2fddb99ee327

**Reply by Dinho4** [reply]
C:/Users/Malesela/Downloads/CIV5000W Simulation Project.inp is my entire path but still with the code snippet to run simulation model provided above, it cannot open file. The following is what I view;
>>> from pyswmm import Simulation
>>>
>>> with Simulation(r'C:/Users/Malesela/Downloads/CIV5000W Simulation Project.inp') as sim:
...     for step in sim:
...         print(sim.current_time)
...

    Cannot open input file C:/Users/Malesela/Downloads/CIV5000W Simulation Project.inpTraceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/root/tutorial_env/lib/python3.9/site-packages/pyswmm/simulation.py", line 69, in __init__
    self._model.swmm_open()
  File "/root/tutorial_env/lib/python3.9/site-packages/pyswmm/swmm5.py", line 217, in swmm_open
    solver.swmm_open(inpfile, rptfile, binfile)
  File "/root/tutorial_env/lib/python3.9/site-packages/swmm/toolkit/solver.py", line 509, in swmm_open
    return _solver.swmm_open(f1, f2, f3)
Exception:
  ERROR 303: cannot open input file.

Is it possible that pyswmm installation does not have access to it?

**Reply by jennwuu** [reply]
@Dinho4 can you double check that you can open the inp file using Python File Open in the same file?
```
with open(r'C:/Users/Malesela/Downloads/CIV5000W Simulation Project.inp') as out:
    for line in out:
        print(line)

```

**Reply by Dinho4** [reply]
Let me start from the bottom, the following is what I view when I try 'pip install pyswmm' from my command prompt:
pip install pyswmm
Collecting pyswmm
  Using cached pyswmm-1.1.1-py3-none-any.whl (95 kB)
Collecting julian==0.14
  Using cached julian-0.14-py3-none-any.whl
Collecting swmm-toolkit==0.8.2
  Using cached swmm-toolkit-0.8.2.zip (623 kB)
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\Malesela\AppData\Local\Temp\pip-install-liq8mmiq\swmm-toolkit_dfd175add43c439691b0d6b692d26f19\setup.py", line 20, in <module>
          from skbuild import setup
      ModuleNotFoundError: No module named 'skbuild'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

**Reply by jennwuu** [reply]
Looks like swmm-toolkit was not installed properly. Are you using python 64 bit?

**Reply by michaeltryby** [reply]
You have swmm-toolkit.zip in your pip cache. It's trying to build locally. 
Try using 
`pip install pyswmm --no-cache-dir` 

This should force it to download the wheel instead.

**Reply by Dinho4** [reply]
> Looks like swmm-toolkit was not installed properly. Are you using python 64 bit?

I am using python 64 bit.

**Reply by Dinho4** [reply]
> You have swmm-toolkit.zip in your pip cache. It's trying to build locally. Try using `pip install pyswmm --no-cache-dir`
> 
> This should force it to download the wheel instead.

I get the following when using 'pip install pyswmm --no-cache-dir:
C:\Users\Malesela>pip install pyswmm --no-cache-dir
Collecting pyswmm
  Downloading pyswmm-1.1.1-py3-none-any.whl (95 kB)
     ---------------------------------------- 95.9/95.9 KB 785.9 kB/s eta 0:00:00
Collecting swmm-toolkit==0.8.2
  Downloading swmm-toolkit-0.8.2.zip (623 kB)
     ---------------------------------------- 623.1/623.1 KB 7.9 MB/s eta 0:00:00
  Preparing metadata (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> [6 lines of output]
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "C:\Users\Malesela\AppData\Local\Temp\pip-install-bb1ioif6\swmm-toolkit_1f88651fbf7d45c1a0175dc7706db03e\setup.py", line 20, in <module>
          from skbuild import setup
      ModuleNotFoundError: No module named 'skbuild'
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.

**Reply by michaeltryby** [reply]
What Platform are you running on? We have Python 3.9 wheels for Windows, Linux and Mac. It looks like you are running Python 3.9 64 bit on Windows but it keeps trying to install from the source distribution instead of the binary distribution.

**Reply by Dinho4** [reply]
I am running python 3.9 (64 bit) on windows

**Reply by jennwuu** [reply]
You can try installing pyswmm from .whl (swmm_toolkit-0.8.2-cp39-cp39-win_amd64.whl) manually: https://pypi.org/project/swmm-toolkit/#files

**Reply by michaeltryby** [reply]
What processor are you running? ARM, AMD, or Intel? Are you running on a Windows tablet?

**Reply by Dinho4** [reply]
> What processor are you running? ARM, AMD, or Intel? Are you running on a Windows tablet?

The processor is Intel(R) Core(TM) i3-6006U CPU @ 2.00GHz   2.00 GHz and I am running on a Windows Laptop.

**Reply by Dinho4** [reply]
> You can try installing pyswmm from .whl (swmm_toolkit-0.8.2-cp39-cp39-win_amd64.whl) manually: https://pypi.org/project/swmm-toolkit/#files

what command should I use in the command prompt when 'installing pysmm from .whl (swmm_toolkit-0.8.2-cp39-cp39-win_amd64.whl) manually' after downloading the file mentioned?

**Reply by jennwuu** [reply]
`pip install swmm_toolkit-0.8.2-cp39-cp39-win_amd64.whl`

**Reply by Dinho4** [reply]
I am using Ubuntu instead of Command prompt now and the following is what I get:
(tutorial_env) (base) root@LAPTOP-LFDK3825:~# pip install pyswmm
Requirement already satisfied: pyswmm in ./tutorial_env/lib/python3.9/site-packages (1.1.1)
Requirement already satisfied: swmm-toolkit==0.8.2 in ./tutorial_env/lib/python3.9/site-packages (from pyswmm) (0.8.2)
Requirement already satisfied: julian==0.14 in ./tutorial_env/lib/python3.9/site-packages (from pyswmm) (0.14)
Requirement already satisfied: aenum==2.2.6 in ./tutorial_env/lib/python3.9/site-packages (from swmm-toolkit==0.8.2->pyswmm) (2.2.6)

But this is what I get when I try to import Simulation:
(tutorial_env) (base) root@LAPTOP-LFDK3825:~# from pyswmm import Simulation
from: can't read /var/mail/pyswmm

**Reply by michaeltryby** [reply]
So I guess binary install swmm_toolkit from wheel on Windows didn't work? Did it?

**Reply by jennwuu** [reply]
@Dinho4 you need to enable python before importing pyswmm.

**Reply by Dinho4** [reply]
> So I guess binary install swmm_toolkit from wheel on Windows didn't work? Did it?

I did not work

**Reply by Dinho4** [reply]
> @Dinho4 you need to enable python before importing pyswmm.

How do you enable it?

**Reply by michaeltryby** [reply]
> > So I guess binary install swmm_toolkit from wheel on Windows didn't work? Did it?
> 
> I did not work

What error did you get?

**Reply by Dinho4** [reply]
C:\Users\Malesela>pip install swmm_toolkit-0.8.2-cp39-cp39-win_amd64.whl
Defaulting to user installation because normal site-packages is not writeable
WARNING: Requirement 'swmm_toolkit-0.8.2-cp39-cp39-win_amd64.whl' looks like a filename, but the file does not exist
ERROR: swmm_toolkit-0.8.2-cp39-cp39-win_amd64.whl is not a supported wheel on this platform.

**Reply by michaeltryby** [reply]
When you invoke python there is a message written before the prompt. What does it say? On my machine I see the following: 
```
(base) michaeltryby@iMac-Pro ci-tools % python
Python 3.9.6 | packaged by conda-forge | (default, Jul 11 2021, 03:36:15) 
[Clang 11.1.0 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```
What do you see?

**Reply by Dinho4** [reply]
(tutorial_env) (base) root@LAPTOP-LFDK3825:~# python
Python 3.9.5 (default, Jun  4 2021, 12:28:51)
[GCC 7.5.0] :: Anaconda, Inc. on linux
Type "help", "copyright", "credits" or "license" for more information.
>>>

**Reply by jennwuu** [reply]
@Dinho4 not to add more confusion, you tried to install .whl in C:\Users\Malesela, what happens if you ran `python` in your command prompt? Can you share the printout from your command prompt as well?

I'm looking for information on your python version.

**Reply by michaeltryby** [reply]
Are you running Ubuntu in Windows Subsystem for Linux? Are you aware that you are running "tutorial_env" environment within your "base" environment? I don't think you meant to do that.

**Reply by michaeltryby** [reply]
Trying to install on Windows and Ubuntu is adding plenty of confusion. Just choose one. Ok?

**Reply by Dinho4** [reply]
> @Dinho4 not to add more confusion, you tried to install .whl in C:\Users\Malesela, what happens if you ran `python` in your command prompt? Can you share the printout from your command prompt as well?
> 
> I'm looking for information on your python version.

Yes, I tried to install .whl in C:\Users\Malesela.

The following is what happens when I run python:
C:\Users\Malesela>python
Python 3.10.2 (tags/v3.10.2:a58ebcc, Jan 17 2022, 14:12:15) [MSC v.1929 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> print("Malesela")
Malesela
>>>

**Reply by Dinho4** [reply]
> Trying to install on Windows and Ubuntu is adding plenty of confusion. Just choose one. Ok?

I will stick to installing on windows

**Reply by jennwuu** [reply]
@michaeltryby pip is not installing because Python is 3.10 on Windows

@Dinho4 can you downgrade your Python version to 3.9 or lower?

**Reply by Dinho4** [reply]
The following is what I get after installing python 3.8 (64 bit):
C:\Users\Malesela>pip install pyswmm
Defaulting to user installation because normal site-packages is not writeable
Collecting pyswmm
  Using cached pyswmm-1.1.1-py3-none-any.whl (95 kB)
Collecting swmm-toolkit==0.8.2
  Downloading swmm_toolkit-0.8.2-cp38-cp38-win_amd64.whl (905 kB)
     |████████████████████████████████| 905 kB 6.8 MB/s
Collecting julian==0.14
  Using cached julian-0.14.zip (2.8 kB)
Collecting aenum==2.2.6
  Using cached aenum-2.2.6-py3-none-any.whl (43 kB)
Using legacy 'setup.py install' for julian, since package 'wheel' is not installed.
Installing collected packages: aenum, swmm-toolkit, julian, pyswmm
    Running setup.py install for julian ... done
Successfully installed aenum-2.2.6 julian-0.14 pyswmm-1.1.1 swmm-toolkit-0.8.2
WARNING: You are using pip version 20.2.3; however, version 22.0.3 is available.
You should consider upgrading via the 'c:\program files\python38\python.exe -m pip install --upgrade pip' command.

**Reply by michaeltryby** [reply]
So it worked

**Comment by bemcdonnell** [comment]
@Dinho4 or maybe something is wrong with the format of your INP file.

**Reply by Dinho4** [reply]
What advice can you give if the file format wrong

**Reply by michaeltryby** [reply]
Try removing spaces from the filename

**Reply by Dinho4** [reply]
I have tried removing the spaces as suggested but still cannot open the input file. I do not why pyswmm cannot import Simulation because my model in EPA SWMM 5.1 runs simulation.

**Reply by Dinho4** [reply]
Thanks a lot @bemcdonnell @michaeltryby and @jennwuu, the simulation is initialized and is running.

**Reply by jennwuu** [reply]
No problem @Dinho4 happy to help out
